### ISSUE TITLE
How configuration would work in .NET

### ISSUE THREAD

**Issue by lmolkova** [open-telemetry/opentelemetry-dotnet#6163]
[.NET configuration](https://learn.microsoft.com/en-us/dotnet/core/extensions/configuration) is a very common and convenient way to configure all sorts of .net apps.
It supports multiple configuration sources, but merges them in one common configuration object. 

The common pattern is to define configuration for all components in a single file - `appsettings.json` and add/override things via other sources. 

Given that it's the idiomatic way to do things in this ecosystem:
- users will be surprised that they need to keep their otel config in a different file than the rest of their configuration.
- they would likely be surprised they need to configure the file with the env var

The ideal outcome as I see it is that configuration story should leave a room for SDKs to decide how onboarding happens and whether it's done within a separate file or as a part of existing configuration system.
I assume there could be a similar ask to integrate otel configuration with spring config, microprofile config or something else.

In .NET it might look like a otel-specific section in the `appsettings.yaml` and no env var to opt into - this is pretty much how it could happen [today](https://github.com/open-telemetry/opentelemetry-dotnet/blob/af57de2b017a09a85b3bd4323d7b01b0174b5742/examples/AspNetCore/appsettings.json#L6), but I assume the schema will need to change to match new one defined in this repo.

**Comment by martincostello** [comment]
@stevejgordon Sounds like there's (a lot of?) overlap with the Declarative Config work here.

**Comment by stevejgordon** [comment]
Thanks for the ping @martincostello - yes, significant overlap.

#7413 introduces the initial declarative config, which takes an "overlay" approach for basic scalar values which integrates with .NET's `IConfiguration` pipeline rather than introducing a separate config system. The YAML file is parsed and projected onto env-var-style keys and registered as a standard `IConfiguration` source, so the SDK picks values up through its existing readers.

Users still need a YAML file as input per the spec requirement, but the `IConfigurationBuilder` integration means it composes naturally with the rest of .NET configuration rather than replacing it.

It's worth noting that .NET users can already configure many component settings (e.g. exporter options) via `appsettings.json` sections using `services.Configure<T>(config.GetSection(...))` - see the [customizing-the-sdk docs](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/customizing-the-sdk/README.md#using-the-net-options-pattern-to-configure-the-sdk) for an example.

So arguably, I think this is largely addressed, unless we want to consider introducing a mechanism to support an "opentelemetry" section in appsettings.json which we automatically read from and bind to the relevant options automatically. That's something I did feel was lacking in the past, although `Configure<T>` was a reasonable workaround. We'd essentially make that automatic for known options types. So for an auto-bind approach, `OtlpExporterOptions` is a good example - auto-binding would essentially just be calling `services.Configure<OtlpExporterOptions>(config.GetSection("OpenTelemetry:Otlp"))` automatically during `AddOtlpExporter()` rather than requiring the user to do it themselves. Auto-binding wouldn't replace declarative config which can express more fine-grained SDK configuration (such as what components (exporters, processors etc.) should be configured in the pipeline), but they could be used independently or compose.

Most .NET developers would be comfortable writing `.AddOtlpExporter()` in code and having the exporter's settings come from appsettings.json automatically. The declarative config use case is primarily zero-code setup (e.g., auto-instrumentation scenarios, ops teams configuring without touching source). These are different audiences.

**Comment by stevejgordon** [comment]
@martincostello Do you think introducing this as a new automatic binding behaviour would be appropriate? If so, I would suggest that I pick this up once declarative config is slightly further along so that I can test there how everything composes. I quite like the idea personally as it's the most idiomatic for developers, especially in simple, smaller, getting started cases.

Our current plan was a layering approach for declarative config which allowably diverges from the spec requirement for it to follow control things. Only some settings may actually work in that flow and that should become a bit clearer soon. We also planned a possible strict mode implementation. If we keep the layered option, we want the application of final values to behave as expected. In reality though, I suspect teams will choose one mechanism or another. Combining them may become confusing. @lmolkova - Does that plan align with your thinking in the original issue?

**Comment by martincostello** [comment]
On paper I think it makes sense for things to compose with the existing configuration system people are used to with modern .NET. If there a specific rough-edges we can discuss those as and when they come up.

**Comment by stevejgordon** [comment]
@martincostello Do you think introducing automatic built-in binding from sections below an OpenTelemetry section in appsettings.json would be appropriate? The only slight risk is for anyone already defining such a section that happens to bind onto our options types which may be surprising. Although, I assume such as section would exist for manual code-based binding already so arguably doesn't break anything. That would make the OOB experience more idiomatic as the default documentation would promote setting those values either in appsettings.json, secrets manager or through any other supported mechanism such as an in-memory provider etc.

**Comment by martincostello** [comment]
I think having it work in the "obvious"/idiomatic way is more valuable than the possibility of breaking the setup of someone who might have happened to have designed their own mapping that matches what we'd be implementing _exactly_. In that case, it would likely end up doing the same thing anyway if the property names were 1:1?

**Comment by stevejgordon** [comment]
Yeah, I would agree. Do you want to assign this to me and I'll pick it up at some through the other configuration work? I'd plan to produce a PR where any finer points can be discussed based on an actual implementation.
